Release/src/http/client/http_client_asio.cpp

```diff
@@ -986,12 +991,31 @@ class asio_context : public request_context, public std::enable_shared_from_this
         }
         if(m_openssl_failed)
         {
-            return verify_cert_chain_platform_specific(verifyCtx, host);
-        }
+
+            if (!is_end_certificate_in_chain(verifyCtx))
+            {
+                // Continue until we get the end certificate.
+                return true;
+            }
+
+            auto chainFunc = [this](const std::shared_ptr<certificate_info>& cert_info) {
+                return m_http_client->client_config().invoke_certificate_chain_callback(cert_info);
+            };
+
+            return http::client::details::verify_cert_chain_platform_specific(verifyCtx, utility::conversions::to_utf8string(host), chainFunc);
+            }
 #endif
 
         boost::asio::ssl::rfc2818_verification rfc2818(host);
-        return rfc2818(preverified, verifyCtx);
+        if(!rfc2818(preverified, verifyCtx))
+        {
+            return false;
+        }
+
+        auto info = std::make_shared<certificate_info>(host, get_X509_cert_chain_encoded_data(verifyCtx));
+        info->verified = true;
+
+        return m_http_client->client_config().invoke_certificate_chain_callback(info);
```

Comment:
On Linux this is called several times. I'm not exactly sure why, but I suppose it is due to the way OpenSSL handles certificate chains (cf. comments in this method). Do we need a similar workaround as for macOS and Android above?

Reply:
Yes, you would need to do the same flow on Linux as per the macOS flow above.

The "is_end_certificate_in_chain" check above should allow the connection until we get the end certificate, once we get that, then we get the OS to build the full cert chain, only after that, then we do the cert-pinning.

If you just wait for the end certificate and don't build the full chain from the OS, you will only have the certs from the SSL handshake, and more than likely that will not be the full chain.